File: sdks/python/windjammer_sdk/sprite.py
"""
Sprite and 2D rendering for Windjammer.

This module provides 2D rendering components and camera.
"""

from typing import Optional
from .math import Vec2, Color
from .ffi import _lib, WjSprite, WjCamera2D, _check_error


class Sprite:
    """2D sprite component."""
    
    def __init__(
        self,
        texture: Optional[str] = None,
        position: Optional[Vec2] = None,
        size: Optional[Vec2] = None,
        color: Optional[Color] = None
    ):
        """
        Create a new sprite.
        
        Args:
            texture: Path to texture file
            position: Position in world space
            size: Size in pixels
            color: Tint color
        """
        self.texture = texture
        self.position = position or Vec2(0, 0)
        self.size = size or Vec2(64, 64)
        self.color = color or Color.white()
        self._handle: Optional[WjSprite] = None
        # TODO: Create sprite with FFI
    
    def __repr__(self) -> str:
        return f"Sprite(texture='{self.texture}', pos={self.position}, size={self.size})"


class Camera2D:
    """2D camera for rendering."""
    
    def __init__(
        self,
        position: Optional[Vec2] = None,
        zoom: float = 1.0,
        rotation: float = 0.0
    ):
        """
        Create a new 2D camera.
        
        Args:
            position: Camera position
            zoom: Zoom level (1.0 = normal)
            rotation: Rotation in radians
        """
        self.position = position or Vec2(0, 0)
        self.zoom = zoom
        self.rotation = rotation
        self._handle: Optional[WjCamera2D] = None
        # TODO: Create camera with FFI
    
    def __repr__(self) -> str:
        return f"Camera2D(pos={self.position}, zoom={self.zoom})"

File: sdks/javascript/src/sprite.ts
/**
 * 2D sprite components.
 */

import { Vec2 } from './math';
import { Transform } from './transform';

/**
 * Options for creating a sprite.
 */
export interface SpriteOptions {
  /** Texture path */
  texture?: string;
  /** Position */
  position?: Vec2;
  /** Size */
  size?: Vec2;
}

/**
 * 2D sprite component.
 */
export class Sprite {
  /** Texture path */
  texture: string;
  
  /** Position */
  position: Vec2;
  
  /** Size */
  size: Vec2;
  
  /** Transform */
  transform: Transform;

  /**
   * Create a new sprite.
   * 
   * @param options - Sprite options
   */
  constructor(options: SpriteOptions = {}) {
    this.texture = options.texture || '';
    this.position = options.position || Vec2.zero();
    this.size = options.size || new Vec2(100, 100);
    this.transform = new Transform();
  }

  toString(): string {
    return `Sprite(texture='${this.texture}', pos=${this.position})`;
  }
}

/**
 * 2D orthographic camera.
 */
export class Camera2D {
  /** Camera position */
  position: Vec2;
  
  /** Zoom level */
  zoom: number;

  /**
   * Create a new 2D camera.
   * 
   * @param position - Camera position
   * @param zoom - Zoom level
   */
  constructor(position: Vec2 = Vec2.zero(), zoom: number = 1.0) {
    this.position = position;
    this.zoom = zoom;
  }

  toString(): string {
    return `Camera2D(pos=${this.position}, zoom=${this.zoom})`;
  }
}

